proposals/js-string-builtins/Overview.md

```diff
@@ -163,6 +163,20 @@ JS Strings are semantically a sequence of 16-bit code units (referred to as char
 
 There is however, the Encoding API for `TextEncoder`/`TextDecoder` which can be used for UTF-8 support. However, this is technically a separate spec from JS and may not be available on all JS engines (in practice it's available widely). This proposal exposes UTF-8 data conversions using this API under separate `wasm:text-encoder` `wasm:text-decoder` interfaces which are available when the host implements these interfaces.
 
+## String constants
+
+String constants may be defined in JS and made available to wasm through a variety of means.
+
+The simplest way is to have a module import each string as an immutable global. This can work for small amounts of strings, but has a high cost for when the number of string constants is very large.
+
+This proposal adds an extension to JS-API compile routine to support optimized 'imported string constants' to address this use-case.
+
+The `WebAssemblyCompileOptions` is extended with a `boolean importedStringConstants` flag. When this is set, the module may define imports of the form `(import "'" "%stringConstant%"" (global externref))`, and the JS-API will use the provided `%stringConstant%` import field name to be the value of the global. This allows for any UTF-8 string to be imported with minimal overhead.
```

Comment:
Would we also want to allow `(ref extern)` as well as `externref`? In practice I think all these would be non-nullable but I'm not sure if that helps engines or not.

Reply:
If we were to choose, I think I would prefer `(ref extern)`  rather than `externref`. This does not force the use of a nullable type. (`ref.as_non_null` is not allowed in constant expressions, so it is not possible to statically cast to a non-nullable type.) Besides, this would be more consistent, since the builtins use the type `(ref extern)` for their return values. 

Allowing `anyref` would make sense as well (this is the type I'm using for JavaScript values so that `ToWebAssembly` convert integers to `i31ref`), but one can use `any.convert_extern` to get this type.

Reply:
Requiring `(ref extern)` instead of `externref` makes sense to me. While engines could infer the non-null-ness, I see no benefit to requiring them to infer that, as opposed to putting it right into the type; on the contrary, using `(ref extern)` would make it a little bit easier to implement a high-performing engine.

Reply:
I agree that allowing `(ref extern)` and `(ref null extern)` (via subtyping) makes sense, this was a bit of an oversight.

Reply:
@eqrion: `(ref extern)` is a subtype of `(ref null extern)`, not the other way round, so it doesn't fall out automatically. We can certainly allow both; but do you see any drawback to specifically requiring non-nullable `(ref extern)`?

Reply:
AFAICS replacing `extern` with `$string` eventually is orthogonal to nullability?

I agree that providing a `(global (ref extern))` to an `(import (global externref))` is fine; the same global can be provided to an `(import (global (ref extern)))`. Once type imports arrive, a `(global (ref $string))` can satisfy an `(import (global (ref $string)))` as well as an old `(import (global (ref extern)))` (if that's even relevant; I suspect it won't be, because both Wasm modules and their JS glue code will be regenerated by new toolchains to make use of new features). I don't see how nullability would help with the transition?

(Also, currently JS code can provide strings directly, without ever calling `new WebAssembly.Global()` explicitly, and I sure hope that that will remain the case when type imports arrive, for efficiency's sake.)